pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,312 @@
+		return
+	}
+
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("failed to get valid endpoint metadata, skipping update to egress policy.")
+		return
+	}
+
+	// Remove old: check if the endpoint was previously selected by any of the policies.
+	if policies, ok := manager.policyEndpoints[epData.id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			err := manager.deleteEgressMap(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error updating endpoint mapping.")
+				return
+			}
+		}
+		delete(manager.policyEndpoints, epData.id)
+		delete(manager.epDataStore, epData.id)
+	}
+
+	// Upsert new: check if current policies select new endpoint. Also updates endpiont cache
+	manager.epDataStore[epData.id] = epData
+	for _, config := range manager.policyConfigs {
+		if config.policyConfigSelectsEndpoint(epData) {
+			err := manager.upsertPolicyEndpoint(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error upserting pod mapping for pod.")
+				return
+			}
+		}
+	}
+}
+
+// OnDeleteEndpoint is the event handler for endpoint deletions.
+func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	var epData *endpointMetadata
+	var err error
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("failed to get valid endpoint metadata, abort deleting endpoint mapping")
+		return
+	}
+
+	if policies, ok := manager.policyEndpoints[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, epData)
+		}
+		delete(manager.policyEndpoints, id)
+	}
+	delete(manager.epDataStore, id)
+}
+
+func (manager *Manager) getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint) (*endpointMetadata, error) {
+	var ipv4s []string
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	for _, pair := range endpoint.Networking.Addressing {
+		if pair.IPV4 != "" {
+			ipv4s = append(ipv4s, pair.IPV4)
+		}
+	}
+
+	data := &endpointMetadata{
+		ips:    ipv4s,
+		labels: labels.NewLabelsFromModel(endpoint.Identity.Labels).K8sStringMap(),
```

Comment:
Same here for the `endpoint.Identity` being nil

Reply:
fixed. thx